File: app/local_plugins/agent_runtime/ios/Classes/pipeline/VadEngine.swift
import Foundation

/// VadEngine — iOS 静音检测引擎
class VadEngine {
    let sessionId: String
    weak var eventSink: AgentEventSink?

    private let speechThresholdDb: Float
    private let silenceDurationMs: TimeInterval
    private var isSpeaking = false
    private var lastSpeechTime = Date.distantPast

    init(
        sessionId: String,
        eventSink: AgentEventSink,
        speechThresholdDb: Float = -40,
        silenceDurationMs: TimeInterval = 0.8
    ) {
        self.sessionId = sessionId
        self.eventSink = eventSink
        self.speechThresholdDb = speechThresholdDb
        self.silenceDurationMs = silenceDurationMs
    }

    /// 处理一帧 PCM 音频（Int16 数据）
    func processFrame(_ pcmData: [Int16], sttNode: SttPipelineNode) {
        let rms = calculateRms(pcmData)
        guard rms > 0 else { return }
        let db = 20 * log10(rms)

        if db > speechThresholdDb {
            lastSpeechTime = Date()
            if !isSpeaking {
                isSpeaking = true
                sttNode.onSttRawEvent(kind: "vadSpeechStart", text: nil, isFinal: false)
            }
        } else {
            if isSpeaking && Date().timeIntervalSince(lastSpeechTime) > silenceDurationMs {
                isSpeaking = false
                sttNode.onSttRawEvent(kind: "vadSpeechEnd", text: nil, isFinal: false)
            }
        }
    }

    private func calculateRms(_ samples: [Int16]) -> Float {
        guard !samples.isEmpty else { return 0 }
        let sum = samples.reduce(0.0) { $0 + Double($1) * Double($1) }
        return Float(sqrt(sum / Double(samples.count)))
    }
}

File: app/local_plugins/agent_runtime/android/src/main/kotlin/com/aiagent/agent_runtime/pipeline/VadEngine.kt
package com.aiagent.agent_runtime.pipeline

import com.aiagent.agent_runtime.*

/**
 * VadEngine — 静音检测引擎（备用，Azure STT 内置 VAD）
 *
 * 当使用 Azure STT SDK 时，SDK 内部已处理语音活动检测（speechStartDetected/speechEndDetected）。
 * 此引擎保留用于将来接入不含 VAD 的 STT 实现。
 */
class VadEngine(
    private val sessionId: String,
    private val eventSink: AgentEventSink,
    private val speechThresholdDb: Float = -40f,
    private val silenceDurationMs: Long = 800L,
) {
    private var isSpeaking = false
    private var lastSpeechTime = 0L

    /**
     * 处理一帧 PCM 音频数据（由录音线程调用）
     * @param pcmFrame 16-bit PCM 数据
     */
    fun processFrame(pcmFrame: ShortArray) {
        val rms = calculateRms(pcmFrame)
        val db = 20 * Math.log10(rms.toDouble()).toFloat()

        if (db > speechThresholdDb) {
            lastSpeechTime = System.currentTimeMillis()
            if (!isSpeaking) {
                isSpeaking = true
                eventSink.onSttEvent(SttEventData(sessionId, requestId = "", kind = "vadSpeechStart"))
            }
        } else {
            if (isSpeaking && (System.currentTimeMillis() - lastSpeechTime) > silenceDurationMs) {
                isSpeaking = false
                eventSink.onSttEvent(SttEventData(sessionId, requestId = "", kind = "vadSpeechEnd"))
            }
        }
    }

    private fun calculateRms(pcm: ShortArray): Float {
        if (pcm.isEmpty()) return 0f
        var sum = 0.0
        for (sample in pcm) sum += sample * sample
        return Math.sqrt(sum / pcm.size).toFloat()
    }

    fun release() {
        isSpeaking = false
    }
}
